AI question generation sample questions

miguelaenlle opened the pull request:
Related to #11451.

- Added interactive example questions to the AI question generation page.

![image](https://github.com/user-attachments/assets/1e96690c-a153-462f-b59e-2c08a1cd6249)

- The sample question accordion begins expanded if the course has no question drafts or AI question generation prompts. 

- Example questions are interactive. Users can grade their responses or generate new variants:

https://github.com/user-attachments/assets/dd4cb646-c514-4364-9d1b-dc9145a3fd42

- AI question generation has been modified to only use one prompt. This change was merged from this PR: #11753 

- The fill prompt button copies the example question's prompt into the prompt input field:

![image](https://github.com/user-attachments/assets/4f362640-83f4-49c0-877e-7ab7f482de10)

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +274,111 @@ export function InstructorAIGenerateDrafts({
+      </div>
+      <div class="card-body">
+        <p id="question-preview-title">
+          Suppose a ball is thrown from a level surface at a [angle]° angle with
```
Let the user specify initial values

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +274,111 @@ export function InstructorAIGenerateDrafts({
+            <span>Dot product = </span>
+          </span>
+          <input
+            id="question-user-response"
```
`question-preview-response`

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
+  const questionUserResponse = document.querySelector('#question-user-response') as HTMLInputElement;
+  const gradeResponse = document.querySelector('#grade-response') as HTMLSpanElement;
+
+  function updateSampleQuestionPreview(id: string) {
```
Move all of the sample question logic into a separate JS file? Or at least a separate function?

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +274,111 @@ export function InstructorAIGenerateDrafts({
+            type="text"
+            class="form-control"
+          />
+          <span id="grade-response" class="input-group-text d-none">
```
`question-preview-grade`

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +274,111 @@ export function InstructorAIGenerateDrafts({
+        >
+          New variant
+        </button>
+        <button id="grade-button" type="button" class="btn btn-primary">
```
question-preview-grade-button

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +274,111 @@ export function InstructorAIGenerateDrafts({
+      </div>
+      <div class="card-footer d-flex justify-content-end">
+        <button
+          id="new-variant-button"
```
question-preview-new-variant-button

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
 import { onDocumentReady } from '@prairielearn/browser-utils';
 
+function handleGrade() {
+  const gradeButton = document.querySelector('#grade-button') as HTMLButtonElement;
```
Redundant: These query selectors are called each time handleGrade is called.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
+  
+  const responseText = questionUserResponse.value?.toString();
+
+  console.log('answer', answer, 'response text', responseText);
```
Remove the console log

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
+  console.log('answer', answer, 'response text', responseText);
+
+  if (responseText === answer) {
+    gradeResponse.className = 'input-group-text';
```
Redundant

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
+  const responseText = questionUserResponse.value?.toString();
+
+  console.log('answer', answer, 'response text', responseText);
+
```
Comment the logic here: 100%, green if the user got the question right, otherwise 0% and red background

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsClient.ts:
```diff
@@ -2,12 +2,164 @@ import { observe } from 'selector-observer';
+
+  function updateSampleQuestionPreview(id: string) {
+    if (id === 'dot-product') {
+      // Generate the two vectors
```
Create a function to better abstract this: `generateSampleQuestionVariant`

Input: `id`

Output: 
- Question title
- Answer

Then, update the answer, className, value

`generateSampleQuestionVariant` can call separate functions for each ID type.

miguelaenlle reviewed (commented):
Notes from my self PR review.

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +226,112 @@ export function InstructorAIGenerateDrafts({
+                      role="tab"
+                      aria-controls="home"
+                      aria-selected="${index === 0 ? 'true' : ''}"
+                      data-id="${examplePrompt.id}"
```
Is this information still necessary? I don't think it is

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +226,112 @@ export function InstructorAIGenerateDrafts({
+  `;
+}
+
+function SampleQuestionPreview(startPrompt: ExamplePrompt) {
```
Preview -> demo

miguelaenlle replied:
Update IDs as previously suggested

miguelaenlle commented on apps/prairielearn/src/lib/aiGeneratedQuestionSamples.ts:
```diff
@@ -0,0 +1,54 @@
+  promptGeneral: string;
+  promptUserInput: string;
+  promptGrading: string;
+  answerName?: string;
```
Comment what this is: Displayed left label that describes that the user is answering, e.g. Dot Product or Velocity

miguelaenlle reviewed (commented):
Notes from my second self-PR.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+onDocumentReady(() => {
+  const questionDemo = document.querySelector('#question-demo') as HTMLDivElement;
+
+  const questionDemoName = questionDemo.querySelector(
```
Move the query selector calls closer to their actual usage?

miguelaenlle replied:
Question demo contianer?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+    return sampleQuestionSelector.querySelector('.active') as HTMLAnchorElement;
+  }
+
+  // Fill the prompts of the selected question to the prompt input fields
```
prompt input field

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  ) {
+    const feedbackContainer =
+      answerType === 'number' || answerType === 'string'
+        ? answerUnitsFeedbackContainer
```
inputResponseFeedbackContainer?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+    }
+  });
+
+  function setGrade(
```
Explain what this function does: updates the feedback container based on the user's response

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+        };
+    }
+
+    // Clear the user free response field
```
input response field

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+    // Update the examples
+    const examplePrompt = examplePrompts.find((examplePrompt) => examplePrompt.id === id);
+    if (examplePrompt) {
+      // Update the question demo text
```
Question demo -> Container heading?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+      );
+
+      if (examplePrompt.answerType === 'number' || examplePrompt.answerType === 'string') {
+        // Display the free response input
```
input response

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+        // Create the response field placeholder text
+        let placeholderText: string = examplePrompt.answerType;
+
+        // Add relative and absolute tolerance if available
```
This should only be for answerType number

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+        }
+
+        // Update the answer units
+        if (examplePrompt.answerUnits) {
```
This should only be for answerType number

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+                <input
+                  class="form-check-input"
+                  type="${examplePrompt.answerType === 'radio' ? 'radio' : 'checkbox'}"
+                  name="multiple-choice-response"
```
Set the name to option-{optionLetter}?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+            }
+          }
+
+          answer.innerHTML += correctAnswersWithLetters.sort().join(', ');
```
Explain why we sort

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  // Generate the initial variant when the page loads
+  generateSampleQuestionVariant(examplePrompts[0].id);
+
+  // Generate a new variant when the new variant button is clicked
```
Move newVariantButton down here

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  });
+
+  previousQuestionButton.addEventListener('click', () => {
+    const selectedTab = getSelectedSampleQuestion();
```
What if we merge the previous and next question button functions tor educe code repetition? They are basically the same with the exception of a - rather than a + and the examplePromptIndex check

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+
+    if (examplePrompt.answerType === 'number') {
+      const response = userInputResponse.value;
+      const rtol = examplePrompt.rtol;
```
Move these into the if statement

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+      const correctValuesSet = new Set(correctValues);
+
+      let numCorrect = 0;
+      for (const option of optionValues) {
```
Explain this grading scheme

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+
+      const percentCorrect = Math.floor((numCorrect / optionValues.length) * 100);
+
+      // The selected and correct values sets must be equal for a correct answer
```
Remove this comment

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+
+  // Randomly select between even and odd
+  const isEven = Math.random() < 0.5;
+  const selectedNumbers = shuffledNumbers.filter((num) => (isEven ? num % 2 === 0 : num % 2 !== 0));
```
Locate the correct answers - merge this with the correctanswer line

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  const numerator = Math.floor(Math.random() * 10) + 1;
+  const denominator = Math.floor(Math.random() * 10) + 2;
+
+  // Randomly generate an angle between 0 and 2pi
```
2 * PI

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  ];
+
+  const randomIndex = Math.floor(Math.random() * european_countries_and_relative_directions.length);
+  const { country1, country2, direction } = european_countries_and_relative_directions[randomIndex];
```
Comment: Randomly select two countries and their correct direction

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  const randomIndex = Math.floor(Math.random() * european_countries_and_relative_directions.length);
+  const { country1, country2, direction } = european_countries_and_relative_directions[randomIndex];
+
+  const correctStatement = Math.random() < 0.5;
```
Nit: Conditionally change the end of the statement without creating both the correct and random incorrect statement

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
```diff
@@ -0,0 +1,815 @@
+  let discriminant = -1;
+
+  // Keep generating random coefficients until we get real roots
+  while (discriminant < 0) {
```
Nit: This could potentially run forever, add a predetermined set of a, b, c in case that occurs

miguelaenlle commented on apps/prairielearn/assets/stylesheets/instructorAiGenerateDrafts.css:
```diff
@@ -0,0 +1,82 @@
+  display: flex;
+}
+
+/* By default, the feedback badges (0%, 100%) are hidden */
```
Put this by the ones for the multiple choice version?

miguelaenlle commented on apps/prairielearn/assets/stylesheets/instructorAiGenerateDrafts.css:
```diff
@@ -0,0 +1,82 @@
+  display: block;
+}
+
+#answer-label > mjx-container {
```
Explain why we do this: in case answer label is in mathjax

miguelaenlle commented on apps/prairielearn/src/ee/lib/ai-question-generation-benchmark.ts:
```diff
@@ -325,8 +325,6 @@ export async function benchmarkAiQuestionGeneration({
         courseId: course.id,
         authnUserId,
         promptGeneral: benchmark.promptGeneral,
-        promptUserInput: benchmark.promptUserInput,
```
Open a separate PR for the changes making question generation run on just one prompt

miguelaenlle commented on apps/prairielearn/src/lib/aiGeneratedQuestionSamples.ts:
```diff
@@ -0,0 +1,111 @@
+  promptGeneral: string;
+  features: string[];
+  answerType: 'number' | 'radio' | 'checkbox' | 'string';
+  /* Describes what the answer of the user is, e.g. dot product or velocity */
```
The question answer is

miguelaenlle reviewed (commented):
Self-review.

miguelaenlle commented:
#11753 is the pull request that implements single-prompt AI question generation. The changes there have been merged into this PR.

nwalters512 commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestionClient.ts:
There's a lot of somewhat-complex DOM manipulation happening here. This might be a good case to reach for [Preact](https://preactjs.com/)? IIRC you've used React quite a bit before. Let me know what you think!

In case you want to take that direction, I'll hold off on reviewing this file for the time being.

miguelaenlle replied:
As per our discussions last week, I reimplemented the sample question demo using Preact.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +193,162 @@ export function InstructorAIGenerateDrafts({
+                    <div class="overflow-auto">
+                      ${examplePrompts.map((a, index) => {
+                        return html`
+                          <a
```
I'm unable to use a keyboard to navigate through the options here, which is an accessibility issue. I believe this is because of the use of `<a>` without `href`. Did you consider using `<button>` here instead?

miguelaenlle replied:
I updated this dropdown menu to use the `DropdownMenu` component in `react-bootstrap`:

https://github.com/PrairieLearn/PrairieLearn/blob/fa50b7af5574f05dcbfeb5b35eedf02d093922e4/apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx#L110-L122

It is now navigable with a keyboard:

https://github.com/user-attachments/assets/d701413f-d9d7-4f29-8a60-38f5d9ab2139

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +193,162 @@ export function InstructorAIGenerateDrafts({
+function SampleQuestionDemo(initialPrompt: ExamplePrompt) {
+  return html`
+    <div id="question-demo-container" class="card shadow">
+      <div class="card-header d-flex align-items-center p-3 gap-3">
```
IMO `p-3` adds too much whitespace, I think relying on the padding provided by `card-header` will be sufficient.

miguelaenlle replied:
I removed the padding around the `card-header`. Here's what it looks like now:

![image](https://github.com/user-attachments/assets/01ff49ba-1746-454c-962d-fa3648c4bd16)

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +193,162 @@ export function InstructorAIGenerateDrafts({
+                  Next
+                </button>
+              </div>
+              <p class="fw-bold mb-1 mt-3">Question features</p>
+              <ul id="feature-list">
+                ${initialPrompt.features.map((feature) => html`<li>${feature}</li>`)}
+              </ul>
```
I'm somewhat inclined to move this below the question view itself, what do you think about that?

miguelaenlle replied:
That sounds good 👍 . I moved the feature list below the question:

![image](https://github.com/user-attachments/assets/99d23ea4-2a58-49c8-b0a9-c4adce372754)

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.html.ts:
```diff
@@ -287,6 +193,162 @@ export function InstructorAIGenerateDrafts({
+                    </div>
+                  </div>
+                </div>
+                <button id="previous-question-button" type="button" class="btn btn-primary">
+                  Previous
+                </button>
+                <button id="next-question-button" type="button" class="btn btn-primary">
+                  Next
+                </button>
```
Things get a little weird on narrow viewports:

<img width="321" alt="Screenshot 2025-04-14 at 10 44 06" src="https://github.com/user-attachments/assets/a1830af8-584f-4808-8c17-361c2f5a581f" />

Consider wrapping the buttons to a new line when the viewport is sufficiently narrow.

miguelaenlle replied:
I implemented this functionality:

![image](https://github.com/user-attachments/assets/79af4ac6-1230-47ef-bfba-f0427edecfe9)

However, I'm having trouble triggering text truncation. Even with `overflow-hidden` and `text-truncate` button content,

```
<DropdownToggle
          as="button"
          type="button"
          style={{ width: '100%' }}
          className="btn dropdown-toggle border border-gray d-flex justify-content-between align-items-center bg-white"
>
           <span class="overflow-hidden text-truncate">{selectedQuestionName}</span>
</DropdownToggle>
```

this occurs if the viewport is small enough:

![image](https://github.com/user-attachments/assets/4ae967bf-d61f-48ab-9c5d-faed6e0aa855)

Would you have any suggestions on how to resolve this issue? I presume it's related to the use of Flexbox in this component, though let me know your thoughts.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,875 @@
+              <AccordionItem eventKey={'0'}> 
+                <AccordionHeader>Example questions and prompts</AccordionHeader>
+                <AccordionBody>
+                    <div style={{width: '100%'}} className="d-flex align-items-center gap-2 mb-3 flex-wrap">
```
Refactor the toggler and features/prompt into separate components?

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,875 @@
+    );
+})
+
+// TODO: Consider moving this to a separate file
```
Move these functions into a separate file

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+  };
+
+  const handleGenerateNewVariant = async () => {
+    // Clear the previous question content
+    setQuestionContent(null);
+    // Ensure the component is rendered before typesetting MathJax
+    // TODO: This results in flickering. Find a better solution.
+    await new Promise((resolve) => requestAnimationFrame(resolve));
```
Storing Preact nodes in state is a little unconventional. You're probably better off computing `questionContent` on the fly on every render. That'll help get you closer to avoiding this flash.

I think the other missing piece is that you need to convince Preact to throw away the DOM of the MathJax node and make a brand new one. I'll leave another comment about that.

miguelaenlle replied:
This and the other suggestions regarding MathJax rendering worked! Here's what it looks like now:

https://github.com/user-attachments/assets/a9c6e218-6026-4a66-b53f-508f6019c553

nwalters512 replied:
Excellent, glad that did the trick!

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+            (part.content.startsWith('$$') && part.content.endsWith('$$')) ||
+            (part.content.startsWith('$') && part.content.endsWith('$'))
+          ) {
+            return <span key={`question-content-${part.index}`}>{part.content}</span>;
```
This is the other piece I commented on above. Ensuring the key changes when the MathJax does ensures we get a brand-new DOM node on which we can perform type-setting.

```suggestion
            return <span key={`${part.index}-${part.content}`}>{part.content}</span>;
```

miguelaenlle replied:
I made this change in conjunction with rendering the question content `span` tags without a state variable:

https://github.com/PrairieLearn/PrairieLearn/blob/a62658cd3aa4443d80fba66dcefe92e716b219fe/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L221-L242

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+    }
+  };
+
+  const handleTypesetMathjax = async () => {
+    // Ensure the component is rendered before typesetting MathJax
+    // TODO: This results in flickering. Find a better solution.
+    await new Promise((resolve) => requestAnimationFrame(resolve));
+    await onMathjaxTypeset();
+  };
+
+  // When a new variant is loaded, typeset the MathJax content.
+  useLayoutEffect(() => {
+    handleTypesetMathjax();
+  }, [variant]);
```
With the other changes, you should be able to get rid of the `requestAnimationFrame` bit.

miguelaenlle replied:
I was able to remove the `requestAnimationFrame` calls after the other changes.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+      // Round the answer to 4 decimal places
+      return Math.round(variant.correctAnswer * 1e4) / 1e4;
+    }
+    {
+      return variant.correctAnswer;
+    }
```
Nit:

```suggestion
    return variant.correctAnswer;
```

nwalters512 replied:
@miguelaenlle looks like this wasn't actually changed; unresolving the conversation for now.

miguelaenlle replied:
Sorry about that. I committed the suggestion 👍

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+  }, [prompt.id]);
+
+  const handleUpdateAnswerLabel = async () => {
+    setAnswerLabel(null);
+
+    // Ensure the component is rendered before typesetting MathJax
+    // TODO: This results in flickering. Find a better solution.
+    await new Promise((resolve) => requestAnimationFrame(resolve));
```
My hope is that with the other changes here, you can stop using an effect +state here and just read off `prompt.answerLabel` directly.

miguelaenlle replied:
Yup, it reads off `prompt.answerLabel` directly now 👍

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+    handleUpdateAnswerLabel();
+  }, [prompt.answerLabel]);
+
+  useEffect(() => {
+    if (answerLabel) {
+      onMathjaxTypeset();
+    }
+  }, [answerLabel]);
```
You probably want `useLayoutEffect` here.

Consider having just a single `useLayoutEffect` hook call with multiple items in its dependencies array.

miguelaenlle replied:
Sounds good; I replaced this with a single `useLayoutEffect` hook:

https://github.com/PrairieLearn/PrairieLearn/blob/a62658cd3aa4443d80fba66dcefe92e716b219fe/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L115-L117

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+    <Card id="question-demo-card" className="shadow">
+      <CardHeader>
+        <div className="d-flex align-items-center gap-2">
+          <p id="question-demo-name" className="mb-0">
```
Are IDs (both here and elsewhere) necessary? If not, let's remove them to keep things clean.

miguelaenlle replied:
These IDs were not necessary. I looked through all the IDs, and removed the unneeded ones/kept the necessary ones. 


The following were necessary:
- `answer-label` (So that `instructorAiGenerateDrafts.css` can remove the margin from MathJax containers)
- `sample-questions` (For rendering the sample question react component) 
- `id={`check-${index}`}` (So that `FormCheck` element labels are clickable)

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,405 @@
+      <FormControl
+        value={userInputResponse}
+        type="text"
+        aria-label="Sample question response"
```
Consider making the above label mandatory and tying the input to that label for improved accessibility.

miguelaenlle replied:
I made this change:

https://github.com/PrairieLearn/PrairieLearn/blob/a62658cd3aa4443d80fba66dcefe92e716b219fe/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L320-L336

I used `aria-describedby` instead of `for` because `InputGroupText` does not have a `for` prop.

nwalters512 replied:
I was actually referring specifically to `aria-label` here; labels and descriptions are different and serve different purposes.

`InputGroupText` renders a span by default, which indeed doesn't have a `for` attribute, but you can use `<InputGroupText as="label" for="...">` to make it render a label instead, which _does_ have a `for` attribute. Then, you can drop both `aria-label` and `aria-describedby` here.

miguelaenlle replied:
Got it, thank you for the info @nwalters512 ! I'll make this change.

miguelaenlle replied:
I implemented this change:

https://github.com/PrairieLearn/PrairieLearn/blob/14822be1db6aabc8911d8d77ecd3bef507584939/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L299-L309

It works as expected with VoiceOver:

![Screenshot 2025-04-30 at 7 15 06 PM](https://github.com/user-attachments/assets/b88b0692-c0d4-48a0-a9d8-7446d174a3a1)

![Screenshot 2025-04-30 at 7 15 03 PM](https://github.com/user-attachments/assets/bd50f8d8-7274-4151-96d4-dc82650c9c0f)

nwalters512 reviewed (commented):
There's some weirdness when switching from a numeric input question to a string input question, the input placeholder isn't reset correctly:

<img width="652" alt="Screenshot 2025-04-25 at 10 13 48" src="https://github.com/user-attachments/assets/4902d1dd-5713-42ca-b559-56e37420150d" />

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,169 @@
+import AccordionOriginal from 'react-bootstrap/Accordion';
+import AccordionBodyOriginal from 'react-bootstrap/AccordionBody';
+import AccordionHeaderOriginal from 'react-bootstrap/AccordionHeader';
+import AccordionItemOriginal from 'react-bootstrap/AccordionItem';
+import ButtonOriginal from 'react-bootstrap/Button';
+import DropdownOriginal from 'react-bootstrap/Dropdown';
+import DropdownItemOriginal from 'react-bootstrap/DropdownItem';
+import DropdownMenuOriginal from 'react-bootstrap/DropdownMenu';
+import DropdownToggleOriginal from 'react-bootstrap/DropdownToggle';
+
+const Accordion = AccordionOriginal as unknown as typeof AccordionOriginal.default;
+const AccordionItem = AccordionItemOriginal as unknown as typeof AccordionItemOriginal.default;
+const AccordionHeader =
+  AccordionHeaderOriginal as unknown as typeof AccordionHeaderOriginal.default;
+const AccordionBody = AccordionBodyOriginal as unknown as typeof AccordionBodyOriginal.default;
+const Button = ButtonOriginal as unknown as typeof ButtonOriginal.default;
+const Dropdown = DropdownOriginal as unknown as typeof DropdownOriginal.default;
+const DropdownToggle = DropdownToggleOriginal as unknown as typeof DropdownToggleOriginal.default;
+const DropdownMenu = DropdownMenuOriginal as unknown as typeof DropdownMenuOriginal.default;
+const DropdownItem = DropdownItemOriginal as unknown as typeof DropdownItemOriginal.default;
```
Got it -- for consistency, I updated the import statements to import directly from `react-bootstrap/cjs/..`

miguelaenlle replied:
I tried importing the React Bootstrap components directly from `react-bootstrap/cjs/...`, but Preact failed to render the components:

![image](https://github.com/user-attachments/assets/14038658-6cfd-4a1b-a8b4-89753ba92d95)

This error occurred:

![image](https://github.com/user-attachments/assets/360a217a-5d08-44fb-9233-a70b0fa95c69)
 
@nwalters512 , would you have any suggestions on how to resolve this issue?

nwalters512 replied:
It might be helpful to `console.log` out the imported components just to see what they are. There also might be some missing pieces from my other open Preact PR, though nothing specifically comes to mind.

nwalters512 replied:
Ah, of course I missed the fact that I was also doing this little `... as unknown as ...` dance in that PR. I think this will be necessary until https://github.com/react-bootstrap/react-bootstrap/issues/6797 is _actually_ fixed (though I think their current strategy of using `exports` in `package.json` to point to faux-ESM is actually going to cause us even more problems than it fixes).

There's nothing to do here for now, for better or worse.

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,169 @@
+            features={selectedQuestion.features.map((feature, index) => ({
+              id: index,
+              text: feature,
+            }))}
```
Agreed; I made this change 👍 

https://github.com/PrairieLearn/PrairieLearn/blob/53cac2cfebe7276569b17bc02d9a458ce64143dc/apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx#L78

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,169 @@
+          {selectedQuestionName}
+        </DropdownToggle>
+        <DropdownMenu>
+          <div>
```
I removed it 👍

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,169 @@
+      <p className="fw-bold mb-1 mt-3">Question features</p>
+      <ul>
+        {features.map((feature) => (
+          <li key={`feature-list-${feature.id}`}>{feature.text}</li>
```
Got it; I edited this loop (and similar ones) to use the index directly as a key. Previously, I received a linter warning that stated not to use the `index` as the key, but this error no longer appears.

nwalters512 replied:
> Previously, I received a linter warning that stated not to use the index as the key, but this error no longer appears.

Ah, interesting, I do actually see that in my editor too, and when I run `make lint-js`. Let's try to avoid linter warnings here, so using something other than `index` would indeed be best. Sorry I forgot about this best practice, my React is clearly a little rusty!

miguelaenlle replied:
All good! I'll use something other than `index` here and in similar locations.

miguelaenlle commented on apps/prairielearn/assets/scripts/tsconfig.json:
```diff
@@ -12,7 +12,17 @@
+    }
   },
-  "include": ["**/*"]
+  "include": ["**/*", "../../src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx"]
```
I removed this change, and it also builds on my end 👍

miguelaenlle commented on apps/prairielearn/src/lib/aiGeneratedQuestionSamples.ts:
I moved `aiGeneratedQuestionSamples.ts` into `src/ee/pages/instructorAiGenerateDrafts/` 👍

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/instructorAiGenerateDrafts.ts:
```diff
@@ -42,7 +43,21 @@ router.get(
+        drafts,
+        // The sample question is initially open if there are no AI question generation drafts
+        // or if the course does not have any AI question generation prompts
+        sampleQuestionOpen: !hasAIQuestionGenerationPrompts || drafts.length === 0,
```
Regarding the last edge case (someone creating a draft locally, than syncing) -- would it be a feasible case in the long term for a user to generate questions with their own OpenAI token, locally? 

I'd assume that would be very rare, so I removed the `hasAIQuestionGenerationPrompts` portion of this statement and the corresponding SQL query. However, let me know your thoughts, @nwalters512 .

nwalters512 replied:
> Regarding the last edge case (someone creating a draft locally, than syncing) -- would it be a feasible case in the long term for a user to generate questions with their own OpenAI token, locally?

It's possibly, but highly unlikely IMO, and I don't think we should try to optimize for it at all. I think simplifying this conditional was the way to go!

miguelaenlle commented on apps/prairielearn/assets/scripts/instructorAiGenerateDraftsSampleQuestions.tsx:
```diff
@@ -0,0 +1,164 @@
+
+  const handleClickPrevious = () => {
+    if (selectedQuestionIndex > 0) {
+      setSelectedQuestionIndex((prevIndex) => prevIndex - 1);
```
Handle this race condition

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+      return;
+    }
+
+    if (variant.answerType === 'number' && prompt.answerType === 'number') {
```
Do we need the prompt.answerType to be number? Is variant.answerTpe === 'number' not enough?

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+      />
+      {(answerUnits || grade !== null) && (
+        <InputGroupText>
+          <span class={grade !== null ? 'me-2' : ''}>{answerUnits}</span>
```
Fix this

miguelaenlle commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/aiGeneratedQuestionSamples.ts:
```diff
@@ -0,0 +1,502 @@
+  | {
+      answerType: 'number' | 'string';
+      /* Numeric tolerances for grading number answers */
+      rtol?: number;
```
Make this available only for answer type number?

miguelaenlle reviewed (commented):
Self review notes.

miguelaenlle commented:
> There's some weirdness when switching from a numeric input question to a string input question, the input placeholder isn't reset correctly:
> 
> <img alt="Screenshot 2025-04-25 at 10 13 48" width="652" src="https://private-user-images.githubusercontent.com/1476544/437543639-4902d1dd-5713-42ca-b559-56e37420150d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU3Njg5MDAsIm5iZiI6MTc0NTc2ODYwMCwicGF0aCI6Ii8xNDc2NTQ0LzQzNzU0MzYzOS00OTAyZDFkZC01NzEzLTQyY2EtYjU1OS01NmUzNzQyMDE1MGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MjdUMTU0MzIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjk2MjFkMDMxMWU1NzNlOTg1Y2VlZjI1OWJiYWFhMWFmOGU5ZjRiYmQwYTlhOTQzOGJjNzBlN2Q5ZWNjY2Y2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.gl_PHX08OLpu3NTG3EcexQ-fgZga5jXG7UboOTnmlAM">

I fixed this issue. If a user switches questions, the placeholder updates correspondingly:

![image](https://github.com/user-attachments/assets/2f99d25a-9144-4012-87ae-64e404cb4319)

![image](https://github.com/user-attachments/assets/c552b0d5-14d8-4da4-8c5c-eac9a8be7d5e)

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,374 @@
+      <CardBody>
+        <p>
+          {
+            // Split the question into different parts: regular text, MathJax, and bold text.
+            // MathJax is delimited by $$...$$ or $...$
+            // Bold text is delimited by **...**.
+            variant ? (
+              variant.question
+                .split(/(\$\$[\s\S]+?\$\$|\$[\s\S]+?\$|\*\*[\s\S]+?\*\*)/g)
+                .filter(Boolean)
+                .map((part, index) => {
+                  // Bold text
+                  if (part.startsWith('**') && part.endsWith('**')) {
+                    return <strong key={`${index}-${part}`}>{part.slice(2, -2)}</strong>;
+                  }
+
+                  // MathJax
+                  if (
+                    (part.startsWith('$$') && part.endsWith('$$')) ||
+                    (part.startsWith('$') && part.endsWith('$'))
+                  ) {
+                    return <span key={`${index}-${part}`}>{part}</span>;
+                  }
+
+                  // Regular text
+                  return <span key={`${index}-${part}`}>{part}</span>;
+                })
```
@miguelaenlle I think this is actually good advice, and it mirrors the new linter warnings.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+  // Used if the question has a checkbox or multiple choice response
+  const [selectedOptions, setSelectedOptions] = useState<Set<string>>(() => new Set<string>());
+
+  const [placeholder, setPlaceholder] = useState('');
```
Rather than storing the placeholder in state, it's probably better to compute it on the fly.

miguelaenlle replied:
I stored `placeholder` as a `const` instead of as a state variable:

https://github.com/PrairieLearn/PrairieLearn/pull/11712/files#diff-c20fbc219773c222f3d5889277712528022277aa0024d0d8f01c11c49370904aR173-R188

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+    });
+  };
+
+  const handleClearSelectedOptions = () => {
+    setSelectedOptions(new Set<string>());
+  };
```
Nit, this function is only used in one place and only does one thing; consider inlining it at the callsite.

miguelaenlle replied:
I made this change:
https://github.com/PrairieLearn/PrairieLearn/blob/36fc4de6f171c47c939576e59bf7ebe2fc6a28fd/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L73-L80

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+            )
+          }
+        </p>
+        {(prompt.answerType === 'number' || prompt.answerType === 'string') && (
+          <NumericOrStringInput
+            userInputResponse={userInputResponse}
+            placeholder={placeholder}
+            grade={grade}
+            answerLabel={prompt.answerLabel}
+            answerUnits={prompt.answerUnits}
+            onChange={setUserInputResponse}
+          />
+        )}
+        {variant && (variant.answerType === 'checkbox' || variant.answerType === 'radio') && (
+          <CheckboxOrRadioInput
+            selectedOptions={selectedOptions}
+            options={variant.options}
+            grade={grade}
+            answerType={variant.answerType}
+            onSelectOption={handleSelectOption}
+          />
+        )}
```
In the first one, we check `prompt.answerType`, in the second, it's `variant.answerType`. Can we maybe standardize on `prompt.answerType`?

miguelaenlle replied:
As per our discussion on Slack, I'll leave this as-is since `answerType` is a discriminant for `variant` and `prompt`, ensuring that `answerLabel/answerUnits` are available fields for `prompt` and `options` is an available field for `variant`.

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+      </CardBody>
+      <CardFooter>
+        <div className="d-flex flex-wrap justify-content-end align-items-center gap-2">
+          <p className="my-0">
+            <i>Answer: {answerText}</i>
+          </p>
```
Is a wrapping `<p>` really necessary here?

nwalters512 commented on apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx:
```diff
@@ -0,0 +1,366 @@
+          <FeedbackBadge grade={grade} />
+        </div>
+      ) : (
+        <></>
```
```suggestion
        null
```

That said: `grade != null && ( ... )` is a more conventional way to do conditional rendering in React.

miguelaenlle replied:
I implemented the suggested change and adjusted this code to use `grade !== null && (....)`:

https://github.com/PrairieLearn/PrairieLearn/blob/7b7bca8f608244862435f8df983c8a4d991cdba8/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L348-L352

I implemented a similar change for the `FeedbackBadge` in `NumericOrStringInput` as well:

https://github.com/PrairieLearn/PrairieLearn/blob/7b7bca8f608244862435f8df983c8a4d991cdba8/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L315

nwalters512 commented:
@miguelaenlle I looked at this with Mariana and we have a few final things to discuss:

- We felt the page looked kind of busy with the example questions expanded by default. I think we should unconditionally start with them collapsed.
- Let's move the text input / button / AI warning above the example questions so that they're collectively the first, most prominent thing on the page.
- We're unsure on the value of the "question features" list. It might be cleaner to remove it? What do you think?
- Mariana was confused by "Fill prompt". She suggested "Use prompt" as an alternative, possibly without the copy icon, and possibly with a short description below the button saying what will happen when it's clicked (or maybe that belongs in a tooltip). Maybe text like "Copy this prompt to the prompt input"?

We're pretty close here, all the broad strokes are correct! Just a few final bits of polish are needed.

miguelaenlle commented:
> * We felt the page looked kind of busy with the example questions expanded by default. I think we should unconditionally start with them collapsed.

Sounds good 👍 . I'll adjust the page to unconditionally begin with the example questions collapsed. 

> * We're unsure on the value of the "question features" list. It might be cleaner to remove it? What do you think?

I agree that the UI looks cleaner without question features, and will remove it. I think it's also not necessary since it's easy to infer what the features of the question are by viewing the sample question. For this sample question, for instance,

![image](https://github.com/user-attachments/assets/2a3015d9-525a-493b-afd6-8571d417fb46)

It's immediately obvious that it's a checkbox input with multiple correct answers. I also assume that users would infer that it is a randomized question (since they can generate new variants). 

> * Mariana was confused by "Fill prompt". She suggested "Use prompt" as an alternative, possibly without the copy icon, and possibly with a short description below the button saying what will happen when it's clicked (or maybe that belongs in a tooltip). Maybe text like "Copy this prompt to the prompt input"?

Sounds good. I'll change "Fill prompt" to "Use prompt" and remove the copy icon. I'll use a tooltip to reduce the amount of text in the Example questions and prompts section.

miguelaenlle commented:
I made the following changes:
- Moved sample questions below the text input/button/AI warning, and set the sample questions to be collapsed by default:

![image](https://github.com/user-attachments/assets/169c57d4-ad6d-46a2-bb80-3d75dfa90d5e)

![image](https://github.com/user-attachments/assets/a224791c-1826-47fa-be75-3ecb284ac8c4)

- Removed the questions features list and associated code/data.

- Changed "Fill prompt" to "Use prompt", and added a tooltip with text "Copy this prompt to the prompt input"

![image](https://github.com/user-attachments/assets/743cb118-c5e7-4b2b-a026-6d8ac3d42e81)

![image](https://github.com/user-attachments/assets/47687b49-48de-43f8-9bd3-3f53480edb61)

nwalters512 reviewed (approved):
LGTM! Thanks for taking care of those final comments. This is ready to merge 🚀
